ecosystem/supersim.md

```diff
@@ -0,0 +1,319 @@
+        ii. Download button/command for the selected configuration
+        iii. Apply button/command
+        iv. Configuration details: show detailed information about the selected configuration
+
+#### Op-simulator interface 
+
+Checks instance status as served by orchestrator interface and listens for RPC calls that require op-simulator intervention. The purpose of this interface is to intercept RPC calls that should go to the interface rather than to anvil instances directly. 
+
+This is essentially a proxy between the app's RPC calls and the op-simulator. 
+
+```mermaid
+flowchart TD
+   subgraph Orchestrator Interface
+       direction LR
+       A[Instance Status] -->|Displays| A1[Running Instances]
+       A -->|Displays| A2[Stopped Instances]
+      
+       B[Event Log] -->|Shows| B1[Recent Events]
+       B -->|Shows| B2[System Messages]
+      
+       C[Mode Selection] -->|Options| C1[Vanilla Mode]
+       C -->|Options| C2[Forking Mode]
+      
+       subgraph Control Panel
+           direction TB
+           D1[CLI Commands] -->|Command Line| D1a[start]
+           D1 -->|Command Line| D1b[stop]
+           D1 -->|Command Line| D1c[restart]
+          
+           D2[UI Buttons] -->|Graphical Interface| D2a[start]
+           D2 -->|Graphical Interface| D2b[stop]
+           D2 -->|Graphical Interface| D2c[restart]
+       end
+   end
+```
+
+```mermaid
+flowchart TD
+  subgraph Op-Simulator Interface
+      direction LR
+
+
+      C -->|any other RPC call| D2[call eth_sendRawTransaction \non Anvil]
+
+
+      C[ETH RPC API] -->|eth_sendRawTransaction| C1[Does TX have \n ExecutingMessage?]
+             C1 -->|no| D2
+
+
+      C1 -->|yes| D1
+
+
+    
+          D1[Check Interop Message Invariants] -->|If Pass|D2
+          D1 -->|If Fail| D3[Return TX failed]
+  
+    
+      D2 --> B3[Return result]
+  end
+```
+
+### Services
+
+Given the application flow, the following services will be part of supersim: 
+
+1. **Orchestrator**: This is the main focus and interface for developers using supersim. 
+    
+    * Runs all of the other services in supersim and manages their lifecycle
+    * Provides an interface to the user
+    
+        * Initially a static interface.
+        * Medium-term, offer an API interface that the orchestrator exposes to make it easier to build new offchain services that hook into the runtime.
+
+2. **Anvil Instances**: Anvil instances are run by the orchestrator after initial prep work to allow them to mirror the Superchain interop environment.
+
+    * Should be run in `--optimism` mode so the extra fields in transactions are there.
+    * Orchestrator should also run an L1 instance, though L1 withdrawals won’t initially be supported.
+    * Eventually supersim can be extended to be used with HardHat and other local simulation tools as well.
+
+3. **Genesis-applier** : The `genesis-applier` takes genesis files output by the monorepo and applies them to anvil instances.
```

Comment:
The monorepo currently outputs `allocs-X.json` files, where `X` is a feature-set like a hardfork. E.g. `allocs-ecotone.json`. We might just want to output an additional `allocs-supersim.json` with any pre-configured things for the supersim, if more internal EVM things have to change.

Also note that some L2 contracts contain references to the L1, which would need to be updated, to make it work e.g. with different L1 deployments, chain-IDs, etc. Any improvements here, to make templating easier, like not using `immutable` but `storage` (so contract bytecode never ha to be touched) would be really helpful, but has not been prioritized yet (even though it's currently affecting the monorepo testing, the superchain-registry, and the supersim).

Reply:
Ooh the allocs approach is interesting! We have been talking about the best ways to identify what exactly should be used for genesis, and this could be a really helpful option. 

We'll look into the L1 references in the L2 contracts. Maybe we can help make the changes if it makes sense. I'd like to be able to use the contracts as is rather than have to edit them on our end, if at all possible.

Reply:
It would be useful to directly use the JSON allocs files to populate `anvil` instances in vanilla mode, would be surprised if they don't already support a CLI flag for this, where you can just pass a path to an allocs file and its applied. Then we can avoid the need to write another tool that converts an allocs file into JSON `anvil_setCode`/`anvil_setStorage` calls. Ideally we reuse as much of our existing tooling as possible here, so the deploy script with `--sig runWithStateDump()` so that the same genesis is created as in the production setup

Reply:
@tynes makes sense! we're using `anvil --init` to pass a pre-made `genesis.json`

in terms of generating the state dump, definitely alligned on trying to use the deploy script as is, but basically copied the logic in the python script in the monorepo https://github.com/ethereum-optimism/supersim/pull/53/files.